File: app/src/main/java/id/ac/uasbungaa/ui/activity/user/Detail200020016UserActivity.java
package id.ac.uasbungaa.ui.activity.user;

import android.app.Activity;
import android.content.ContentValues;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Bitmap.CompressFormat;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import id.ac.uasbungaa.R;
import id.ac.uasbungaa.data.local.DatabaseContract;
import id.ac.uasbungaa.data.local.UserHelper;
import id.ac.uasbungaa.data.model.User;
import id.ac.uasbungaa.databinding.ActivityDetail200020016UserBinding;
import id.ac.uasbungaa.helper.Keys;
import id.ac.uasbungaa.helper.ToastHelper;

import java.io.ByteArrayOutputStream;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class Detail200020016UserActivity extends AppCompatActivity {
    private ActivityDetail200020016UserBinding binding;
    private User user;
    private String path = "";
    private UserHelper userHelper;
    private View overlay;
    @NotNull
    public static final String EXTRA_USER = "EXTRA_USER";

    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityDetail200020016UserBinding.inflate(this.getLayoutInflater());
        this.setContentView(binding.getRoot());
        this.overlay = findViewById(R.id.progbar_layout);

        this.user = this.getIntent().getParcelableExtra(EXTRA_USER);
        this.userHelper = new UserHelper(this);

        ActionBar supportActionBar = this.getSupportActionBar();
        if (supportActionBar != null) {
            supportActionBar.setTitle(user.getUsername());
            supportActionBar.setDisplayHomeAsUpEnabled(true);
        }

        binding.etPassword.setText(user.getPassword());
        if (user.getThumbnail() != "") {
            binding.civThumbnail.setImageURI(Uri.parse(user.getThumbnail()));
        } else {
            binding.civThumbnail.setImageResource(R.drawable.user);
        }

        binding.civThumbnail.setOnClickListener((OnClickListener) (new OnClickListener() {
            public final void onClick(View it) {
                checkStoragePermission();
            }
        }));

        binding.btnSubmit.setOnClickListener((OnClickListener) (new OnClickListener() {
            public final void onClick(View it) {
                overlay.setVisibility(View.VISIBLE);
                handleSubmit();
            }
        }));
    }

    private final void handleSubmit() {
        String password = binding.etPassword.getText().toString();
        updateRecord(password);
        finish();
    }

    private final void updateRecord(String password) {
        userHelper.open();
        ContentValues values = new ContentValues();
        values.put(DatabaseContract.USER_200020016.PASSWORD, password);
        if (!path.equals("")) {
            values.put(DatabaseContract.USER_200020016.THUMBNAIL, path);
        }
        userHelper.update(String.valueOf(user.getId()), values);
    }

    private final void checkStoragePermission() {
        if (this.checkSelfPermission("android.permission.WRITE_EXTERNAL_STORAGE") != PackageManager.PERMISSION_GRANTED) {
            this.requestPermissions(new String[]{"android.permission.WRITE_EXTERNAL_STORAGE"}, Keys.WRITE_EXTERNAL_CODE);
        } else {
            this.checkCameraPermission();
        }
    }

    private final void checkCameraPermission() {
        if (this.checkSelfPermission("android.permission.CAMERA") != PackageManager.PERMISSION_GRANTED) {
            this.requestPermissions(new String[]{"android.permission.CAMERA"}, Keys.CAMERA_PERMISSION_CODE);
        } else {
            Intent intent = new Intent("android.media.action.IMAGE_CAPTURE");
            this.startActivityForResult(intent, Keys.CAMERA_REQUEST);
        }
    }

    public void onRequestPermissionsResult(int requestCode, @NotNull String[] permissions, @NotNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == Keys.CAMERA_PERMISSION_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ToastHelper.INSTANCE.showToast("Camera Permission Granted", this);
                Intent intent = new Intent("android.media.action.IMAGE_CAPTURE");
                this.startActivityForResult(intent, Keys.CAMERA_REQUEST);
            } else {
                ToastHelper.INSTANCE.showToast("Camera Permission Denied", this);
            }
        }

        if (requestCode == Keys.WRITE_EXTERNAL_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ToastHelper.INSTANCE.showToast("Storage Permission Granted", this);
                this.checkCameraPermission();
            } else {
                ToastHelper.INSTANCE.showToast("Storage Permission Denied", this);
            }
        }

    }

    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == Keys.CAMERA_REQUEST && resultCode == Activity.RESULT_OK) {
            Bitmap photo = (Bitmap) data.getExtras().get("data");
            photo.compress(
                    CompressFormat.JPEG,
                    100,
                    new ByteArrayOutputStream()
            );
            path = MediaStore.Images.Media.insertImage(
                    this.getContentResolver(),
                    photo,
                    user.getUsername() + "_Image",
                    null
            );
            binding.civThumbnail.setImageURI(Uri.parse(path));
        }
    }

    public boolean onOptionsItemSelected(@NotNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.btn_delete:
                this.deleteUser();
                break;
            case android.R.id.home:
                this.finish();
        }

        return super.onOptionsItemSelected(item);
    }

    private final void deleteUser() {
        userHelper.open();
        userHelper.delete(String.valueOf(user.getId()));
        userHelper.close();
        finish();
    }

    public boolean onCreateOptionsMenu(@Nullable Menu menu) {
        this.getMenuInflater().inflate(R.menu.menu_user, menu);
        return super.onCreateOptionsMenu(menu);
    }
}

File: app/src/main/java/id/ac/uasbunga/ui/activity/Detail200020016UserActivity.kt
package id.ac.uasbunga.ui.activity

import android.Manifest
import android.app.Activity
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.MediaStore
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.lifecycleScope
import id.ac.uasbunga.R
import id.ac.uasbunga.data.local.DatabaseContract.USER_200020016.Companion.PASSWORD
import id.ac.uasbunga.data.local.DatabaseContract.USER_200020016.Companion.THUMBNAIL
import id.ac.uasbunga.data.local.UserHelper
import id.ac.uasbunga.data.model.User
import id.ac.uasbunga.databinding.ActivityDetail200020016UserBinding
import id.ac.uasbunga.helper.Keys.Companion.CAMERA_PERMISSION_CODE
import id.ac.uasbunga.helper.Keys.Companion.CAMERA_REQUEST
import id.ac.uasbunga.helper.Keys.Companion.WRITE_EXTERNAL_CODE
import id.ac.uasbunga.helper.ToastHelper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream
import java.io.File

class Detail200020016UserActivity : AppCompatActivity() {
    companion object {
        const val EXTRA_USER = "EXTRA_USER"
    }

    private lateinit var binding: ActivityDetail200020016UserBinding
    private lateinit var user: User
    private var path: String = ""
    private lateinit var userHelper: UserHelper
    private lateinit var overlay: View

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityDetail200020016UserBinding.inflate(layoutInflater)
        setContentView(binding.root)
        overlay = findViewById(R.id.progbar_layout)

        user = intent.getParcelableExtra(EXTRA_USER)!!
        userHelper = UserHelper.getInstance(this)

        supportActionBar?.title = user.username
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        binding.etPassword.setText(user.password)
        if (user.thumbnail != "") {
            binding.civThumbnail.setImageURI(Uri.parse(user.thumbnail))
        } else {
            binding.civThumbnail.setImageResource(R.drawable.user)
        }

        binding.civThumbnail.setOnClickListener {
            checkStoragePermission()
        }

        binding.btnSubmit.setOnClickListener {
            overlay.visibility = View.VISIBLE
            handleSubmit()
        }
    }

    private fun handleSubmit() {
        val password = binding.etPassword.text.toString()
        updateRecord(password)
        finish()
    }

    private fun updateRecord(password: String) {
        userHelper.open()
        val values = ContentValues()
        values.put(PASSWORD, password)
        if (path != "") {
            values.put(THUMBNAIL, path)
        }
        userHelper.update(user.id.toString(), values)
        userHelper.close()
    }

    private fun checkStoragePermission() {
        if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                WRITE_EXTERNAL_CODE
            )
        } else {
            checkCameraPermission()
        }
    }

    private fun checkCameraPermission() {
        if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.CAMERA), CAMERA_PERMISSION_CODE)
        } else {
            val intent = Intent(android.provider.MediaStore.ACTION_IMAGE_CAPTURE)
            startActivityForResult(intent, CAMERA_REQUEST)
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == CAMERA_PERMISSION_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ToastHelper.showToast("Camera Permission Granted", this)
                val intent = Intent(android.provider.MediaStore.ACTION_IMAGE_CAPTURE)
                startActivityForResult(intent, CAMERA_REQUEST)
            } else {
                ToastHelper.showToast("Camera Permission Denied", this)
            }
        }
        if (requestCode == WRITE_EXTERNAL_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ToastHelper.showToast("Storage Permission Granted", this)
                checkCameraPermission()
            } else {
                ToastHelper.showToast("Storage Permission Denied", this)
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == CAMERA_REQUEST && resultCode == Activity.RESULT_OK) {
            val photo = data?.extras?.get("data") as Bitmap
            photo.compress(Bitmap.CompressFormat.JPEG, 100, ByteArrayOutputStream())
            path =
                MediaStore.Images.Media.insertImage(
                    this.contentResolver,
                    photo,
                    "${user.username}_Image",
                    null
                )
            binding.civThumbnail.setImageURI(Uri.parse(path))
        }
    }

    private fun storeToGallery(photoPath: String, context: Context) {
        lifecycleScope.launch(Dispatchers.IO) {
            val mediaScanIntent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE)
            val imageFile = File(photoPath)
            val imageUri = Uri.fromFile(imageFile)
            mediaScanIntent.data = imageUri
            context.sendBroadcast(mediaScanIntent)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.btn_delete -> deleteUser()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun deleteUser() {
        userHelper.open()
        userHelper.delete(user.id.toString())
        userHelper.close()
        finish()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_user, menu)
        return super.onCreateOptionsMenu(menu)
    }
}